src/uct/tcp/tcp_ep.c

```diff
@@ -1442,16 +1456,21 @@ uct_tcp_ep_am_prepare(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
         return status;
     }
 
-    ucs_assertv(ep->tx.buf == NULL, "ep=%p", ep);
-
-    ep->tx.buf = ucs_mpool_get_inline(&iface->tx_mpool);
-    if (ucs_unlikely(ep->tx.buf == NULL)) {
+    status = uct_tcp_ep_ctx_buf_alloc(ep, &ep->tx, &iface->tx_mpool);
+    if (ucs_unlikely(status != UCS_OK)) {
         goto err_no_res;
     }
 
     *hdr          = ep->tx.buf;
     (*hdr)->am_id = am_id;
 
+    if (tx_sn_inc && (++ep->tx.sn == ep->last_acked_sn)) {
+        /* If the TX sequence number is now the same as the last acked sequence
+         * number, ensure that they are different to request ACK through PUT in
+         * TCP ep flush operation */
+        --ep->last_acked_sn;
+    }
```

Comment:
maybe this PR could be done simpler (and w/o adding more counters):
1. increase SN for flush operations as well
2. keep flag on the ep of "whether there was put without flush":
   - put turns the flag on
   - flush is nop is flag is off, otherwise - put and turn the flag off